sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.jena/src/com/ge/research/sadl/jena/JenaBasedSadlModelProcessor.java

```diff
@@ -4236,6 +4261,130 @@ public Object processExpression(ValueTable expr) throws InvalidNameException, In
+			}
+			else {
+				addError("Unexpected null right-hand side of binary operation", first);
+			}
+			first = ((BinaryOperation)first).getLeft();
+		}
+		
+		Object cls = processExpression(first);
+		List<Object> restObjs = new ArrayList<Object>();
+		for (Expression rexpr : restExpressions) {
+			restObjs.add(processExpression(rexpr));
+		}
+		NamedNode thereExistsIdentity = null;
+		if (cls instanceof VariableNode) {
+			thereExistsIdentity = (VariableNode) cls;
+			cls = ((VariableNode)cls).getType();
+		}
+		else {
+			if (cls instanceof Object[] && 
+					((Object[])cls)[0] instanceof NamedNode &&
+					((Object[])cls)[1] instanceof GraphPatternElement) {
+				if (((Object[])cls)[1] instanceof TripleElement) {
+					restObjs.add(0, ((Object[])cls)[1]);
+				}
+				else if (((Object[])cls)[1] instanceof Junction) {
+					List<GraphPatternElement> expanded = IntermediateFormTranslator.junctionToList((Junction)((Object[])cls)[1]);
+					int insertAt = 0;
+					for (GraphPatternElement gpe : expanded) {
+						if (gpe instanceof TripleElement) {
+							restObjs.add(insertAt++, gpe);
+						}
+						else {
+							addError("Expected a triple pattern but found '" + gpe.toString(), matchExpr);
+						}
+					}
+				}
+				cls = ((Object[])cls)[0];
+				thereExistsIdentity = (NamedNode) cls;
+				if (cls instanceof VariableNode) {
+					cls = ((VariableNode)cls).getType();
+				}
+			}
+			else {
+				addError("Did not find expected beginning of there exists.", first);
+			}
+		}
+		Expression plusExpr = expr.getPlus();
+		Object plusObj = null;
+		if (plusExpr != null) {
+			plusObj = processExpression(plusExpr);
+		}
+		BuiltinElement tebi = new BuiltinElement();
+		tebi.setFuncName("thereExists");
+		tebi.addArgument(nodeCheck(cls));
+		for (Object robj : restObjs) {
+			TripleElement rtr = null;
+			if (robj instanceof TripleElement) {
+				rtr = (TripleElement) robj;
+			}
+			else if (robj instanceof Object[] && ((Object[])robj)[1] instanceof TripleElement) {
+				rtr = (TripleElement) ((Object[])robj)[1];
+			}
+			else {
+				addError("failed to find match triple", matchExpr);
+			}
+			if (rtr != null) {
+				if (rtr.getSubject().equals(thereExistsIdentity)) {
+					tebi.addArgument(rtr.getPredicate());
+					tebi.addArgument(rtr.getObject());
+				}
+				else if (rtr.getObject() != null && rtr.getObject().equals(thereExistsIdentity)){
+					tebi.addArgument(rtr.getSubject());
+					tebi.addArgument(rtr.getPredicate());
+				}
+				else {
+					String msg ="It appears that there is a triple in 'there exists' that does not reference the matching instance.";
+					if (!useArticlesInValidation) {
+						msg += " Are you using articles in translation?";
+					}
+					msg += " Does the triple belong in the 'plus' portion?";
+					addError(msg, rtr.getContext() != null ? (EObject)rtr.getContext() : expr);
+				}
+			}
+		}
+		if (plusObj != null) {
+			NamedNode plusNode = new NamedNode(SadlConstants.SADL_IMPLICIT_MODEL_URI + "#Plus");
```

Comment:
The plus object may be a list and need to loop through them as for the match objects?

Reply:
Fix is committed. Please review.